### PR TITLE
cylc-rose-rose-stem

### PR DESCRIPTION
**Pull request by wxtim**
This pair of PRs 
- [cylc/cylc-flow #4116](https://github.com/cylc/cylc-flow/pull/4116) 
- [cylc-rose #45](https://github.com/cylc/cylc-rose/pull/45) 

close [metomi/rose #2448](https://github.com/metomi/rose/issues/2448)

**This is a change which will require a `pip install -e .[all]` to test - there are entry point changes.**

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).

- [x] Appropriate functional tests are included in the [cylc-rose #45](https://github.com/cylc/cylc-rose/pull/45). These are a direct translation of the tests that can be found in the [existing rose-stem test](https://github.com/metomi/rose/blob/master/t/rose-stem/00-run-basic.t).
- [x] No unit tests have been added - this is deliberate - since this is supposed to be a direct translation this would be out of scope.
- [x] No change log entry has been added to Cylc because this feature is only available if the Cylc-Rose plugin is installed.
- [x] Cylc Rose doesn't have a Change log. It probably should.

- [x] I have documented that this feature exists in the cylc-rose ``__init__`` file. Further documentation, or translation of existing Rose-stem docs is required, but IMO out of scope.
- [x] No dependency changes.

### CODE REVIEW

File: cylc/rose/stem.py [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+    """Implement rose stem."""
+    # use the cylc install option parser
+    parser = get_option_parser()
+
+    # TODO: add any rose stem specific CLI args that might exist
+    # On inspection of rose/lib/python/rose/opt_parse.py it turns out that
+    # opts.group is stored by the --task option.
+    parser.add_option(
+        "--task", "--group",
+        help="Specifies a source tree to include in a suite.",
```
Comment (datamel) [inline cylc/rose/stem.py:471]:
```suggestion
        help="Specifies a source tree to include in a workflow.",
```

File: cylc/rose/stem.py [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+        action="append",
+        metavar="PATH/TO/FLOW",
+        default=[],
+        dest="stem_groups")
+    parser.add_option(
+        "--source",
+        help="Specifies a source tree to include in a suite.",
```
Comment (datamel) [inline cylc/rose/stem.py:478]:
```suggestion
        help="Specifies a source tree to include in a workflow.",
```

**Review by datamel — COMMENTED** [review]
I have a failing test on this branch  in `tests/test_functional_post_install.py` however this is also failing on master for me.
I have noticed throughout `suite.rc` is used, would it be an idea to convert these to flow.cylc.
Otherwise, from my reading of the original and new code, I think the functionality has been preserved from rose stem.

File: tests/functional/test_rose_stem.py [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+    (basetemp / 'conf/keyword.cfg').write_text(
+        f"location{{primary}}[foo.x]={url}"
+    )
+    monkeymodule.setenv('FCM_CONF_PATH', str(confdir))
+    # Check out a working copy of the repo:
+    suitename = workingcopy.parts[-1]
+    subprocess.run(split(f'fcm checkout -q fcm:foo.x_tr {workingcopy}'))
+    # Copy suite into working copy.
+    test_src_dir = Path(__file__).parent / '12_rose_stem'
+    for file in ['rose-suite.conf', 'suite.rc']:
```
[the remark this answers is not in the document]
Comment (wxtim) [reply tests/functional/test_rose_stem.py:157]: I'll do this manually, bc I need to create the actually file too.

File: cylc/rose/stem.py [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+from metomi.rose.host_select import HostSelector
+from metomi.rose.popen import RosePopener
+from metomi.rose.reporter import Reporter, Event
+from metomi.rose.resource import ResourceLocator
+from metomi.rose.suite_run import SuiteRunner
+
+
+DEFAULT_TEST_DIR = 'rose-stem'
+ROSE_STEM_VERSION = 1
+SUITE_RC_PREFIX = '[jinja2:suite.rc]'
```
[the remark this answers is not in the document]
Comment (wxtim) [reply cylc/rose/stem.py:42]: I'm trying to preserve old code. For now...

File: cylc/rose/stem.py [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+
+
+class NameSetEvent(Event):
+
+    """Event to report a name for the suite being set."""
+
+    LEVEL = Event.V
+
+    def __repr__(self):
+        return "Suite is named %s" % (self.args[0])
```
[the remark this answers is not in the document]
Comment (wxtim) [reply cylc/rose/stem.py:76]: Again, I'm leaving old code as is, if you don't mind.

File: cylc/rose/stem.py (hunk 1 above)
Comment (wxtim) [reply cylc/rose/stem.py:471]: I think that this remains suite.

**Comment by wxtim** [comment]
@datamel  - sorry about the pushback, but I'm trying not to touch the code I moved more than strictly necessary to get it working. Besides, I think rose-stem-suites will remain rose-stem-suites.

**Review by oliver-sanders — COMMENTED** [review]
Code looks good, still need to test.

**Comment by oliver-sanders** [comment]
Lets aim to get this into the beta-0 meta package if we can since it is so close.

**Review by datamel — APPROVED** [review]
Tests now passing for me. Thanks @wxtim.

**Review by oliver-sanders — APPROVED** [review]

**Pull request merged**